src/IW/Sync/Search.hs

```diff
@@ -28,70 +35,117 @@ import qualified Data.Text as T
 import qualified Data.Vector as V
 
 
--- | Fetch all repositories built with the Haskell language.
-fetchAllHaskellRepos
-    :: ( MonadIO m
+githubSearch
+    :: forall a env m.
+       ( MonadIO m
        , WithError m
        , WithLog env m
+       , FromJSON a
+       , Typeable a
        )
-    => m [GitHub.Repo]
-fetchAllHaskellRepos = liftGitHubSearchToApp searchHaskellRepos
+    => [Text] -- ^ Paths
+    -> Text   -- ^ Query String
+    -> Day    -- ^ From day
+    -> Day    -- ^ To day
+    -> Int    -- ^ Page
+    -> m [a]
+githubSearch paths queryString from to page = liftIO (executeRequest' $ query paths queries) >>= \case
+    Left err -> throwError $ githubErrToAppErr err
+    Right (SearchResult count vec) -> do
+        log Info $ "Query executed with the following paths and query parameters: " <> show paths <> " " <> show queries
+        log Info $ "Fetched total of " <> show count <> " " <> typeName @a <> "s..."
+        searchLimit <- getSearchRateLimit
+        log Info $ "Search rate limit information: " <> show searchLimit
+        pure $ V.toList vec
+  where
+    queries :: [(ByteString, Maybe ByteString)]
+    queries = pagination <> [("q", Just $ encodeUtf8 $ strip queryString <> " " <> dateRange from to)]
```

Comment:
I feel like it will help readability if you introduce helper functions to construct queries. Again, this function seems to do a lot. So it makes to split into separate functions. For example, you can create `buildGitHubQuery` function that takes query string, page number and returns result of a `query` function. 

P.S. I would use `GitHub.query` function qualifed because `query` is too generic name and we have `query` in DB functions, so fully qualifying will help.